### ISSUE TITLE
Spotless messing up multi-line TODOs

### ISSUE THREAD

**Issue by Zabuzard** [Together-Java/TJ-Bot#233]
### Overview

When writing a multi-line TODO/FIXME/NOTE, following lines usually have a single whitespace prefix to indicate that the TODO continues on the next line.

IDEs such as Intellij and Eclipse recognize this and adjust their highlighting, as well as their TODO-viewer tools.

However, Spotless removes the prefix whitespace. Leading to IDEs thinking that the following lines are comments that are not part of the TODO anymore.

### Example

Suppose a TODO like
```java
// TODO Lorem
//  ipsum
```
Note the extra space in front of `ipsum`, i.e. `// _ipsum`.

When running Spotless, it changes it to
```java
// TODO Lorem
// ipsum
```
removing the extra space, i.e. `// ipsum`.

See the following screenshots that show the state before and after applying Spotless:
![before](https://user-images.githubusercontent.com/13614011/139020159-22331481-51f0-4b63-80f0-bb6f524be04e.png)
![after](https://user-images.githubusercontent.com/13614011/139020163-3358ef58-16b8-4652-964e-dfe49a72d436.png)
In particular, pay attention to the purple highlighting.

**Comment by marko-radosavljevic** [comment]
@Zabuzard
I researched this a bit. 

tl;dr Elegant solution doesn't exist, and it's probably not worth our resources to try to fix it.

Note that this is IntelliJ IDEA feature (introduced in 2018.3). Eclipse does not support it, and I assume other IDEs as well.

### Some backstory and debugging info might be useful in the future:

* * *

Spotless is more like a facade for multiple formatters, packed together in a neat tool that works with many languages, and many build tools. Actual formatting is done by the tools like eclipse formatter, google java formatter, prettier.

We use eclipse formatter, with a custom configuration based on [eclipse java google style](https://github.com/google/styleguide/blob/gh-pages/eclipse-java-google-style.xml).

<details> 
  <summary> Debugging time: (click) </summary>

With that in mind, let's change to  google java format:

```diff
diff --git a/build.gradle b/build.gradle
index b52d9b9..4256027 100644
--- a/build.gradle
+++ b/build.gradle
@@ -65,7 +65,7 @@ subprojects {
         java {
             // Excludes build folder since it contains generated java classes.
             targetExclude("build/**")
-            eclipse('4.19.0').configFile("${rootProject.rootDir}/meta/formatting/google-style-eclipse.xml")
+            googleJavaFormat()
         }
     }
 }
diff --git a/gradle.properties b/gradle.properties
new file mode 100644
index 0000000..4257286
--- /dev/null
+++ b/gradle.properties
@@ -0,0 +1,5 @@
+org.gradle.jvmargs=--add-exports jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED \
+  --add-exports jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED \
+  --add-exports jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED \
+  --add-exports jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED \
+  --add-exports jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED
 ```
The issue was not reproducible with google format, so the issue must be eclipse format, or our config for eclipse format.
```diff
diff --git a/build.gradle b/build.gradle
index b52d9b9..815821e 100644
--- a/build.gradle
+++ b/build.gradle
@@ -65,7 +65,7 @@ subprojects {
         java {
             // Excludes build folder since it contains generated java classes.
             targetExclude("build/**")
-            eclipse('4.19.0').configFile("${rootProject.rootDir}/meta/formatting/google-style-eclipse.xml")
+            eclipse('4.19.0')
         }
     }
 }
```
(Note that we had to add `--add-exports` jvm args, because google format doesn't play well with jdk16+ encapsulated internals.)

The issue is reproducible once again, so it's the eclipse format, and not our config. Let's try the latest eclipse version.

```diff
diff --git a/build.gradle b/build.gradle
index b52d9b9..b1967fc 100644
--- a/build.gradle
+++ b/build.gradle
@@ -65,7 +65,7 @@ subprojects {
         java {
             // Excludes build folder since it contains generated java classes.
             targetExclude("build/**")
-            eclipse('4.19.0').configFile("${rootProject.rootDir}/meta/formatting/google-style-eclipse.xml")
+            eclipse()
         }
     }
 }
```
Issue reproducible with the latest eclipse formatter. 
</details>

Result: eclipse formatter doesn't support this intelliJ feature, google java format does.

### Possible solutions

* * *

#### Solution 1: `@formatter:off` tag 

Eclipse supports `@formatter:off` tag that disables formatting until `@formatter:on` tag is encountered.

Since we have it enabled in our config,

```xml
<setting id="org.eclipse.jdt.core.formatter.use_on_off_tags" value="true"/>
```

We can do something like this:

```java
// @formatter:off
// TODO Lorem
//  ipsum
// @formatter:on
 ```

Reference: https://archive.eclipse.org/eclipse/downloads/drops/R-3.6-201006080911/eclipse-news-part2.html#JavaFormatter

(Note that this feature can be used for any other formatting issue we have.)

I would argue this is not worth the extra effort, and cost of maintaining it, since we might want to switch formatters.

#### Solution 2: */-
Eclipse supports java comment code conventions, so we can do something like this to disable formatting for this comment block:

```
/*-
* TODO Lorem ipsum2
*  ipsum
*/
```

Reference: https://www.oracle.com/java/technologies/javase/codeconventions-comments.html#350

I would argue this is not worth the extra effort, and it's an unorthodox way of writing todos.

#### Solution 3: disable formatting on comments

I would argue this is the worst option, and definitely not worth it, since, obviously, we would have unformatted comments.

* * *

Note that with all 3 solutions, intelliJ highlighting works, and todo view shows multi-line todo. Eclipse remains unaffected by the introduced changes. 

<details><summary> Images (click) </summary>

IntelliJ:
![image](https://user-images.githubusercontent.com/30927961/142710527-5334f093-f05e-4793-86eb-a59cd776916b.png)

Eclipse:
![image](https://user-images.githubusercontent.com/30927961/142710539-4cacebe7-58d8-4f94-a781-b54daf750fe9.png)
</details>

Also, solution list is not conclusive. It's possible to change formatters. But that would require bigger rework of the formatting, which would inevitably bring new issues.

### Conclusion 

I'm labelling this as wontfix, we can reevaluate this decision later, in case we decide changing formatters is worthwhile.

**Comment by Zabuzard** [comment]
I agree, thanks for the extensive research and the valuable background info! Wont fix, closed.
